nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        inputs = inputs.transpose(1, 2)
+
+        inputs = inputs * inputs_masks
+        out = inputs.unsqueeze(3)
+        out = self.conv(out)
+        out = out.view(out.shape[0], out.shape[1], -1).contiguous()
+        self.gru.flatten_parameters()
+        _, out = self.gru(out)
+
+        return out.squeeze(0)
+
+
+class StyleAttention(NeuralModule):
+    def __init__(self, gru_hidden=128, gst_size=128, n_style_token=10, n_style_attn_head=4):
+        super(StyleAttention, self).__init__()
+        self.input_size = gru_hidden
+        self.output_size = gst_size
+        self.n_token = n_style_token
+        self.n_head = n_style_attn_head
+        self.token_size = self.output_size // self.n_head
+
+        self.tokens = torch.nn.Parameter(torch.FloatTensor(self.n_token, self.token_size))
+
+        self.q_linear = torch.nn.Linear(self.input_size, self.output_size)
+        self.k_linear = torch.nn.Linear(self.token_size, self.output_size)
+        self.v_linear = torch.nn.Linear(self.token_size, self.output_size)
+
+        self.tanh = torch.nn.Tanh()
+        self.softmax = torch.nn.Softmax(dim=2)
+        self.temperature = (self.output_size // self.n_head) ** 0.5
+        torch.nn.init.normal_(self.tokens)
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D'), EncodedRepresentation()),
+            "token_id": NeuralType(('B'), Index(), optional=True),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
+        q = self.q_linear(inputs.unsqueeze(1))
+        k = self.k_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+        v = self.v_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+
+        q = q.view(bs, q.shape[1], self.n_head, self.token_size)
+        k = k.view(bs, k.shape[1], self.n_head, self.token_size)
+        v = v.view(bs, v.shape[1], self.n_head, self.token_size)
+
+        q = q.permute(2, 0, 1, 3).contiguous().view(-1, q.shape[1], q.shape[3])
+        k = k.permute(2, 0, 3, 1).contiguous().view(-1, k.shape[3], k.shape[1])
+        v = v.permute(2, 0, 1, 3).contiguous().view(-1, v.shape[1], v.shape[3])
+
+        scores = torch.bmm(q, k) / self.temperature
+        scores = self.softmax(scores)
+
+        style_emb = torch.bmm(scores, v).squeeze(1)
+        style_emb = style_emb.contiguous().view(self.n_head, bs, self.token_size)
+        style_emb = style_emb.permute(1, 0, 2).contiguous().view(bs, -1)
+
+        return style_emb
+
+
+class SVEmbeddingLinearProjectionNetwork(torch.nn.Module):
+    def __init__(self, speaker_embedding_dim, symbols_embedding_dim):
+        super(SVEmbeddingLinearProjectionNetwork, self).__init__()
+        self.speaker_proj = torch.nn.Linear(speaker_embedding_dim, symbols_embedding_dim)
+        self.act_fn = torch.nn.ReLU()
```

Comment:
Should we have a more generic LinearActivation class which takes (input_dim, output_dim, activation)?

Reply:
Is it a good idea to use a ReLU activation on a residual? Other architectures like ResNet and transformer can do it because they apply batch norm or layer norm afterwards.

Reply:
We don't use the activation function, so I remove it.